mypy/checkexpr.py

```diff
@@ -1130,22 +1135,15 @@ def check_overload_call(self,
         unioned_result = None  # type: Optional[Tuple[Type, Type]]
         unioned_errors = None  # type: Optional[MessageBuilder]
         union_success = False
-        if any(isinstance(arg, UnionType) and len(arg.relevant_items()) > 1  # "real" union
-               for arg in arg_types):
-            erased_targets = self.overload_erased_call_targets(plausible_targets, arg_types,
-                                                               arg_kinds, arg_names, context)
-            unioned_callable = self.union_overload_matches(erased_targets)
-
-            if unioned_callable is not None:
-                unioned_errors = arg_messages.clean_copy()
-                unioned_result = self.check_call(unioned_callable, args, arg_kinds,
-                                                 context, arg_names,
-                                                 arg_messages=unioned_errors,
-                                                 callable_name=callable_name,
-                                                 object_type=object_type)
-                # Record if we succeeded. Next we need to see if maybe normal procedure
-                # gives a narrower type.
-                union_success = unioned_result is not None and not unioned_errors.is_errors()
+        if any(self.real_union(arg) for arg in arg_types):
+            unioned_errors = arg_messages.clean_copy()
+            unioned_result = self.union_overload_result(plausible_targets, args, arg_types,
+                                                        arg_kinds, arg_names,
+                                                        callable_name, object_type,
+                                                        context, arg_messages=unioned_errors)
+            # Record if we succeeded. Next we need to see if maybe normal procedure
+            # gives a narrower type.
+            union_success = unioned_result is not None and not unioned_errors.is_errors()
 
         # Step 3: We try checking each branch one-by-one.
         inferred_result = self.infer_overload_return_type(plausible_targets, args, arg_types,
```

Comment:
Github won't let me leave a comment down below after doing the simplification suggested up above, I think we can also simplify the following if statements to something roughly like:

    # Success! Stop early by returning the best among normal and unioned.
    if inferred_result is not None and unioned_result is not None:
        return inferred_result if is_subtype(inferred_result[0], unioned_result[0]) else unioned_result
    elif inferred_result is not None:
        return inferred_result
    elif unioned_result is not None:
        return unioned_result